azure/scope/managedmachinepool.go

```diff
@@ -0,0 +1,250 @@
+func NewManagedMachinePoolScope(ctx context.Context, params ManagedMachinePoolScopeParams) (*ManagedMachinePoolScope, error) {
+	_, _, done := tele.StartSpanWithLogger(ctx, "scope.NewManagedMachinePoolScope")
+	defer done()
+
+	if params.Cluster == nil {
+		return nil, errors.New("failed to generate new scope from nil Cluster")
+	}
+
+	if params.ControlPlane == nil {
+		return nil, errors.New("failed to generate new scope from nil ControlPlane")
+	}
+
+	helper, err := patch.NewHelper(params.InfraMachinePool, params.Client)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to init patch helper")
+	}
+
+	return &ManagedMachinePoolScope{
+		Client:               params.Client,
+		Cluster:              params.Cluster,
+		ControlPlane:         params.ControlPlane,
+		MachinePool:          params.MachinePool,
+		InfraMachinePool:     params.InfraMachinePool,
+		patchHelper:          helper,
+		ManagedClusterScoper: params.ManagedControlPlaneScope,
+	}, nil
+}
+
+// ManagedMachinePoolScope defines the basic context for an actuator to operate upon.
+type ManagedMachinePoolScope struct {
+	Client      client.Client
+	patchHelper *patch.Helper
+
+	azure.ManagedClusterScoper
+	Cluster          *clusterv1.Cluster
+	MachinePool      *expv1.MachinePool
+	ControlPlane     *infrav1exp.AzureManagedControlPlane
+	InfraMachinePool *infrav1exp.AzureManagedMachinePool
+}
+
+// PatchObject persists the cluster configuration and status.
+func (s *ManagedMachinePoolScope) PatchObject(ctx context.Context) error {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "scope.ManagedMachinePoolScope.PatchObject")
+	defer done()
+
+	conditions.SetSummary(s.InfraMachinePool)
+
+	return s.patchHelper.Patch(
+		ctx,
+		s.InfraMachinePool,
+		patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+			clusterv1.ReadyCondition,
+		}})
+}
+
+// Close closes the current scope persisting the cluster configuration and status.
+func (s *ManagedMachinePoolScope) Close(ctx context.Context) error {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "scope.ManagedMachinePoolScope.Close")
+	defer done()
+
+	return s.PatchObject(ctx)
+}
+
+func (s *ManagedMachinePoolScope) AgentPoolAnnotations() map[string]string {
+	return s.InfraMachinePool.Annotations
+}
+
+// AgentPoolSpec returns an azure.AgentPoolSpec for currently reconciled AzureManagedMachinePool.
+func (s *ManagedMachinePoolScope) AgentPoolSpec() azure.AgentPoolSpec {
+	return buildAgentPoolSpec(s.ControlPlane, s.MachinePool, s.InfraMachinePool)
+}
+
+func buildAgentPoolSpec(managedControlPlane *infrav1exp.AzureManagedControlPlane,
+	machinePool *expv1.MachinePool,
+	managedMachinePool *infrav1exp.AzureManagedMachinePool) azure.AgentPoolSpec {
+	var normalizedVersion *string
+	if machinePool.Spec.Template.Spec.Version != nil {
+		v := strings.TrimPrefix(*machinePool.Spec.Template.Spec.Version, "v")
+		normalizedVersion = &v
+	}
+
+	replicas := int32(1)
+	if machinePool.Spec.Replicas != nil {
```

Comment:
What situation would we have `machinePool.Spec.Replicas` be nil? And if so, is there any particular reason why we want to set the default replicas to 1?

Reply:
this (and the entire file) was copied directly as-is from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2168/files#diff-cfd94adbc2ff767742e457e64836e93e88c6e03c3a64c04619c54b7ed76f284eL574